Private lib schema

pivotaljohn opened the pull request:
As part of this PR we decided on some UX decisions described below:
* When `--enable-experiement-schema` flag is **not** present, and a schema file is provided, a warning is shown. **New in this PR** the relative path to the file is also shown. (Consider the case where root library has a schema, and a private library has a schema, there will be **two** warning messages now).
* Rules around how the `--enable-experiement-schema` behaves: 
When the `--enable-experiement-schema` flag is present, and a schema file is given in the root or in a private library, it is used as the base data values and type checked.
When `--enable-experiement-schema` flag is **not** enabled, then if a schema file appears anywhere, the schema is IGNORED and a warning is given. This is because the schema flag is protecting the user against non-GA code.


⚠️   Update based on feedback ⚠️  
We decided to split this original PR into three separate ones.
```
develop 
|_overlay-position       <- PR to address position changes within overlay module (separate for this PR entirely)
|_ refactor-dvpp         <- PR to address refactoring done in data_values_pre_processing.go (must merge prior to this PR)
    |_ private-lib-schema <- this PR
```

cari-lynn committed:
Add test for schema in private library

cari-lynn committed:
Use schema as data values only in library `eval()`

- Does not yet type check schema
- Does not respect `--schema-enabled` flag

Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Don't error when no root lib schema or dv with schema enabled

Commit:
Type check library schema against programmatic data values

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

jtigger committed:
Remove nullable annotation check from MapType

* ytt does not place annotations on maps

Signed-off-by: Steven Locke <slocke@pivotal.io>

jtigger committed:
Reorganize Data Values pre-processing

- ensure that every overlay of data values (regardless of source) is
  checked against schema (if enabled).
- inline ApplyWithDefaults()
- Provide direct error messages by reporting schema errors instead of overlay errors when possible

Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

jtigger committed:
(refactor) tweak some naming / definition ordering

- DataValues.HasLib() => DataValues.HasLibRef()
  (without interpretion, conveys that data values has a
   reference to a library)
- Schema.AsDataValues() => Schema.DefaultDataValues()
  (slightly more revealing of intented use)
- in data_values_pre_processing.go, sorted some function
  definitions in order of usage.
  (eases reading when order of usage matches order of
   definition)

jtigger committed:
Rename Schema tests to improve readability

Renaming:
- use snake_case for test names to significantly improve readability.
  (such methods are *never* invoked)
- use [slightly] more verbose sentence structure to expose relevant
  details, in a fluid expression, remaining frugal with words.
- reword sub-test names to be phrases extending the containing test
  name as a sentence.

Also:
- delete two (2) redundant test cases
- reorder tests and sub-tests like an outline of a feature spec.

cari-lynn committed:
Library functions validate their data values against their schema

Signed-off-by: John Ryan <jtigger@infosysengr.com>

cari-lynn committed:
Respect `--enable-experiment-schema` for libraries

- Test that when schema is disabled, schema is not used

Signed-off-by: John Ryan <jtigger@infosysengr.com>

cari-lynn committed:
Move schema post-processing into Schemas function

cari-lynn committed:
Add path in `--enable-experiment-schema` warning

Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

pivotaljohn commented:
> When --enable-experiement-schema flag is not enabled, then if a schema file appears anywhere, the schema is IGNORED and a warning is given. This is because the schema flag is protecting the user against non-GA code.

I really like this choice. It gives the author the ability to support _both_ modes. Include both a `schema.yml` and `values/default.yml` with basically the same content. If I enable schemas, I get the schema checks. If I don't I get the overlay checks.

cppforlife commented on pkg/yttlibrary/overlay/array.go:
```diff
@@ -43,6 +43,8 @@ func (o Op) mergeArrayItem(
 		if replace {
 			leftArray.Items[leftIdx].Value = newItem.Value
+			//todo: should we be setting the position elsewhere?
+			leftArray.Items[leftIdx].Position = newItem.Position
```
can we keep position related improvements separate from this PR. it's a bit hard to tell if these are the only places where we need to change this -- would love to see holistic evaluation within overlay module.

pivotaljohn replied:
... yup and that PR would precede this one — this fix addresses an actual bug: the wrong file is attributed a value.

cari-lynn replied:
These position changes are now necessary to report the position of the overlay in the schema error message, if it's acceptable, we want to keep the change made to the position of arrays and maps here in [this](https://github.com/vmware-tanzu/carvel-ytt/pull/337) PR, and then address this more comprehensively in the [Overlay Positions](https://github.com/vmware-tanzu/carvel-ytt/pull/341) PR.

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -121,36 +120,20 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
 	libraryLoader := libraryExecutionFactory.New(libraryCtx)
 
-	schemaDocs, err := libraryLoader.Schemas()
+	currSchema, err := libraryLoader.Schemas()
```
is it worth renaming libraryLoader -> rootLibraryLoader

cppforlife commented on pkg/schema/schema.go:
```diff
@@ -73,10 +73,6 @@ func NewMapType(m *yamlmeta.Map) (*MapType, error) {
 		}
 		mapType.Items = append(mapType.Items, mapItemType)
 	}
-	annotations := template.NewAnnotations(m)
```
note for self: removed because it's dead code.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
-		if err != nil {
-			return nil, nil, fmt.Errorf("Templating file '%s': %s", fileInLib.File.RelativePath(), err)
-		}
+	_, checkSchema := o.loader.schema.(*schema.DocumentSchema)
```
something to think about... why do we need a boolean here? why cant we just tell schema to type check against itself. all schemas should be able to do it succesfully.

cppforlife replied:
accessing private data from loader got me curious... it seems that template loader does not use schema at all. it seems weird for template loader to know anything about schema since it's not dealing with data values. i would have expected DataValuesPreProcessing to be initialized with appropriate schema.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
 
-		for _, valuesDoc := range valuesDocs {
-			dv, err := NewDataValues(valuesDoc)
+	allDvs, err := o.collectDataValuesDocs(files)
```
nice little refactor!

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+
+	// merge all Data Values YAML documents into one
+	var dvsForOtherLibraries []*DataValues
+	var dataValuesDoc *yamlmeta.Document
```
i wonder if naming this from dataValuesDoc -> resultDVsDoc would make it a bit more readable. (it's the result from this loop)

pivotaljohn replied:
... and taking that suggestion a step further, if we use the same naming scheme across identifiers: `(qualifier)(noun)`, the eye parses identifiers using the same pattern matcher, reducing cognitive load.

Example:
- `allDvs`
- `otherLibraryDvs`
- `resultDvDoc`
- others?

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+	var dataValuesDoc *yamlmeta.Document
+	for _, dv := range allDvs {
+		switch {
+		case dv.HasLibRef():
```
this switch statement could be made more readable as two if statements. note that type checking below switch statement would happen even if dvs are for a library ref with current impl.

```
if lib dvs {
   ...
   continue
}

if nil {
  ...
} else {
  ...
}
```

the fact that there is 2 if statements leads the reader to see that it's two unrelated pieces of functionality.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+		case dv.HasLibRef():
+			dvsForOtherLibraries = append(dvsForOtherLibraries, dv)
+		case dataValuesDoc == nil:
+			err := o.loader.schema.ValidateWithValues(1)
```
note to self: function goes away as soon as we kill experimental flag.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+			dataValuesDoc = dv.Doc
+		default:
+			if checkSchema {
+				o.annotateWithDefaultMissingOk(dv.Doc)
```
let's pull this out as a separate PR. by adding missing_ok=True in certain cases we are affecting how overlay feature behaves under certain conditions. that's not desirable because we want users to know that same piece of code behaves same way regardless where it's being called from.

quick example that comes to mind: 
```
#@overlay/match by=overlay.subset({"name":"foo"})
- blah: something
```
will behave differently since since overlay module by default sets expects=1.

to start resolving this situtation, i think we need to figure out what happens in various interactions between schema and overlay. what errors would one expect to see when new keys are used, what errors do we expect to see when new keys are used with any type, etc.. let's have a document enumerating these cases and discuss them.

pivotaljohn replied:
... and to clarify, it's not so much that we'd pull this out as a separate PR... but that we'd _either_:

- address this problem _within_ this PR, or,
- temporarily accept this intermediate state where it generally works but has undesired side-effects in some cases... but then _immediately_ address this problem in a subsequent PR.  (consider this change set "large enough" and address this as a separate unit of work)

But either way, we can't remove the `--enable-experiment-schemas` flag until we ensure that this works exact the way we want it.

cari-lynn replied:
We pulled out this to a separate PR where it is addressed, @cppforlife ready for review. The PR is #337 and is a prerequisite to this work so that PR must be merged in prior to this one.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+}
 
-	return dv, libraryValues, nil
+func (o DataValuesPreProcessing) typeAndCheck(dataValuesDoc *yamlmeta.Document) (chk yamlmeta.TypeCheck) {
```
we typically discourage use of named returns. dont think it's necessary in this case.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
+	}
+
+	typeCheck := dataValuesDoc.Check()
+	chk.Violations = append(chk.Violations, typeCheck.Violations...)
```
no need to append to chk.Violations since it's guranteed to be 0.

cppforlife commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -151,30 +178,23 @@ func (o DataValuesPreProcessing) overlay(valuesDoc, newValuesDoc *yamlmeta.Docum
-			return nil, fmt.Errorf("Overlaying additional data values on top of "+
-				"data values from files (marked as @data/values): %s", err)
-		}
+func (o DataValuesPreProcessing) NewEmptyDataValuesDocument() *yamlmeta.Document {
```
seems like this method could be inlined where it's used, or at least made private.

pivotaljohn replied:
... in part because there's no need for any code outside this module to generate an empty Data Values.

cppforlife commented on pkg/workspace/library_loader.go:
```diff
@@ -46,30 +46,36 @@ func NewLibraryLoader(libraryCtx LibraryExecutionContext,
 }
 
-func (ll *LibraryLoader) Schemas() ([]*yamlmeta.Document, error) {
+func (ll *LibraryLoader) Schemas() (Schema, error) {
```
if we are returning a single schema should this method be named Schema() as well?

cppforlife commented on pkg/workspace/library_loader.go:
```diff
@@ -46,30 +46,36 @@ func NewLibraryLoader(libraryCtx LibraryExecutionContext,
-		docs, _, err := DocExtractor{resultDocSet}.Extract(AnnotationSchemaMatch)
-		if err != nil {
-			return nil, err
+			docs, _, err := DocExtractor{resultDocSet}.Extract(AnnotationSchemaMatch)
```
if we get multiple schema docs i would expect that we would blow up for now since we do not allow merging schemas.

cari-lynn replied:
Correct, we have a story coming up to allow multiple schema docs #294. Prior to this story we would also blow up if we get multiple docs, so we haven't added guards for that case.

cppforlife commented on pkg/workspace/library_module.go:
```diff
@@ -318,7 +317,8 @@ func (l *libraryValue) libraryValues(ll *LibraryLoader) (*DataValues, []*DataVal
 	}
 
-	dvs, foundChildDVss, err := ll.Values(append(dvss, afterLibModDVss...), &schema.AnySchema{})
+	schema, err := ll.Schemas()
```
missing err check.

cari-lynn replied:
😨

cppforlife commented on pkg/workspace/library_module.go:
```diff
@@ -318,7 +317,8 @@ func (l *libraryValue) libraryValues(ll *LibraryLoader) (*DataValues, []*DataVal
 
-	dvs, foundChildDVss, err := ll.Values(append(dvss, afterLibModDVss...), &schema.AnySchema{})
+	schema, err := ll.Schemas()
+	dvs, foundChildDVss, err := ll.Values(append(dvss, afterLibModDVss...), schema)
```
it seems that we are using ll.Schemas() and then passing it to ll.Values(). i would argue Values() shouldnt require schema as input and should just use ll.Schemas() internally.

cari-lynn committed:
move schema checking out of template

Signed-off-by: John Ryan <jtigger@infosysengr.com># Please enter the commit message for your changes. Lines starting

cari-lynn committed:
Add test for schema only validates current library

cari-lynn committed:
Rename variables and functions

- Rename Schemas() to Schema()
- Rename variables to use qualifier - noun naming scheme

cari-lynn committed:
Refactor Values() to call Schema()

- dvpp has Schema instead of templateLoader

gcheadle-vmware committed:
Address feedback in private-lib-schema

- move logic for schema assinging type and dvs checking themselves into typeAndCheck()
- change swtich statement in dvpp to if statements
- make newEmptyDataValuesDocument() private

Signed-off-by: Cari Dean <cdean@pivotal.io>

cari-lynn committed:
Remove typechecking on failed data value overlays

- In favor of a more complete solution, we are removing this typecheck
  to be addressed in a later issue.
- Skipping tests where the overlay fails, but we want a schema error.

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>
